provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -106,51 +111,40 @@ type CreateVirtualMachineParams struct {
 	EnableDiskUUID bool
 }
 
-// CreateVirtualMachine creates and powers on a new VM.
-//
-// This method imports an OVF template using the vSphere API. This process
-// comprises the following steps:
-//   1. Ensure the VMDK contained within the OVA archive (args.OVA) is
-//      stored in the datastore, in this controller's cache. If it is
-//      there already, we use it; otherwise we remove any existing VMDK
-//      for the same series, and upload the new one.
-//   2. Call CreateImportSpec [0] with a pre-canned OVF, which validates
-//      the OVF descriptor against the hardware supported by the host system.
-//      If the validation succeeds,/the method returns an ImportSpec to use
-//      for importing the virtual machine.
-//   3. Prepare all necessary parameters (CPU, memory, root disk, etc.), and
-//      call the ImportVApp method [0]. This method is responsible for actually
-//      creating the VM. This VM is temporary, and used only to convert the
-//      VMDK file into a disk type file.
-//   4. Clone the temporary VM from step 3, to create the VM we will associate
-//      with the Juju machine.
-//   5. If the user specified a root-disk constraint, extend the VMDK if its
-//      capacity is less than the specified constraint.
-//   6. Power on the virtual machine.
-//
-// [0] https://www.vmware.com/support/developer/vc-sdk/visdk41pubs/ApiReference/
-// [1] https://www.vmware.com/support/developer/vc-sdk/visdk41pubs/ApiReference/vim.HttpNfcLease.html
-func (c *Client) CreateVirtualMachine(
+// vmTemplateName returns the well-known name to
+// the template VM for this controller and its models
+func vmTemplateName(args CreateVirtualMachineParams) string {
+	return "juju-template-" + args.OVASHA256
+}
+
+// vmTemplatePath returns the a path inside the vSphere datastore
+// where the template VM is housed.
+func vmTemplatePath(args CreateVirtualMachineParams) string {
+	return path.Join(args.VMDKDirectory, args.Series)
+}
+
+// ensureTemplateVM returns a vSphere template VM
+// that instances can be created from.
+func (c *Client) ensureTemplateVM(
 	ctx context.Context,
 	args CreateVirtualMachineParams,
-) (_ *mo.VirtualMachine, resultErr error) {
-
-	// Locate the folder in which to create the VM.
+) (*object.VirtualMachine, error) {
```

Comment:
You need the error return named, otherwise the use of `err` in the deferred func to abort the lease won't necessarily correspond to what's being returned - it'll be the local variable `err`, rather than the return error. For example, the return on line 253:
```go
    return nil, errors.New("SHA-256 hash mismatch for OVA")
```
That doesn't set `err`, so the func will see `err == nil` and not abort the lease, but if the name `err` in the func refers to the named return it will see it.

Basically you need to replace the err local var with the named return parameter err (which the code in the method can still assign to in the same way). Hopefully that makes sense? (This is about the fiddliest that scoping in Go gets.)

Reply:
Good catch, I think I get it.

Reply:
Switched to using named return values, which should resolve this issue.